### ISSUE TITLE
Mongo Shards Support

### ISSUE THREAD

**Issue by battle-mage** [rwynn/monstache#36]
Hello. We are looking for a tool to synchronize our mongoDB to Elastic. But we have sharded collections and we are using MongoS.
We have a workaround that we use in branch(mongos-support) at our fork(kolesa-team) now (https://github.com/kolesa-team/monstache/tree/mongos-support), but it's quite dirty.

Do you plan adding MongoS support to Monstache?

**Comment by rwynn** [comment]
Hi, thanks for trying monstache.  I'll take a look at adding this.  I think I will need to add code at a fundamental level (in the gtm library that monstache uses) to support this cleanly.

**Comment by rwynn** [comment]
Just added a commit that adds the `enable-shards` config option.  Please let me know how that goes.  Much of the change went into the gtm library used by monstache.

**Comment by rwynn** [comment]
Just added a few more changes.  You should use the `config` server url for the monstache connection URL and then make sure that `enable-shards` is turn on.  Use the `verbose` config to make sure it's picking up the shards.

**Comment by battle-mage** [comment]
Thank you for the quick reaction! We will try it out next week, then I will give a feedback.

**Comment by rwynn** [comment]
Just realized I still have some work to do as the logic is a bit off. Currently it will work for the most part but monstache state updates and queries inside JavaScript or Golang plugins will go against the mongo config server session instead of to the mongos session.

I will need to refactor such that mongo-url is the URL to mongos and a new arg mongo-config-url is the URL to the config servers. 

Also, I plan to try to listen for new shards and add those dynamically. But not till after I fix the issue above.

**Comment by rwynn** [comment]
I've refactored now to the point where I think it makes sense.  I've removed the `enable-shards` option.  Now, when working with a sharded cluster you would run as follows:

```
monstache --verbose --mongo-url mongosHost:27017 --mongo-config-url configHost:27018
```
The config server URL is a trigger to retrieve the shards and start syncing on all of them.  Writes of monstache state data and reads inside Javascript or Golang plugins occur on the mongos connection.

**Comment by rwynn** [comment]
@battle-mage I've just added a commit that also listens for new shards added the cluster.  Previously, monstache would only read the shards once.  Now it also tails the config server to discover and incorporate shards when they are added to the cluster so that a restart of monstache is not necessary.

**Comment by battle-mage** [comment]
Hello, sorry for the late response, we weren't able to test it for a while.
Thank you for your work, looks like monstache now supports shards accordingly, but we faced another problem, when we do a lot of operations in Mongo in a short period of time.
We faced next situation:
When Monstache is configured to keep more than 1 connection to ElasticSearch (elasticsearch-max-conns is NOT 1)
1) a document being created, (doc version 1)
2) a document is being updated (doc version 2)

If theese two operations are executed concurrently in parallel connection - second http-request might complete BEFORE the first one is completed. In this case ACTUAL document with current real date is being replaced by the OLD one. So we have old data in Elastic without seeing any errors.

If we set elasticsearch-max-conns to 1 - everything works fine.

Don't know what to do in this situation, maybe we should always include the EXACT timestamp from oplog and ignore update on Elastic like in this article - https://discuss.elastic.co/t/conditional-update-to-the-document/64964/2

Or maybe there can be another solution?

P.S.: ofcourse, we can limit concurrent elastic connections to one, but perhaps it will significally decreace the performance.

**Comment by rwynn** [comment]
@battle-mage  thanks for getting back and the new issue report.  I'll close this sharding issue and create another one for the concurrency issue.  I think it should be solvable via the `version` attribute and external versioning via `version_type`.  It will look a little strange as the version numbers will be MongoDB timestamps from the oplog which are 64 bit numbers.  But it should solve the issue and reject the create if it comes in after the update during high concurrency due to max-conns greater than 1.
